[Fix] graph support 0-Size tensor

grybd opened the pull request:
跑下面这段最小复现代码,就可以复现 issue https://github.com/Oneflow-Inc/oneflow/issues/6958 提到的问题
```python
import unittest
from oneflow.test_utils.automated_test_util import *
import oneflow as flow
import oneflow.unittest


class TestReLUModule(flow.unittest.TestCase):
    @autotest(auto_backward=False)
    def test_relu_module_with_0shape_data(test_case):
        m = torch.nn.ReLU()
        m.train(random())
        device = random_device()
        m.to(device)
        x = random_pytorch_tensor(4, 2, 3, 0, 3).to(device)
        y = m(x)
        return y
    
if __name__ == "__main__":
    unittest.main()
```

grybd commented:
**在跑test_activation.py测试脚本时把跑def test_relu_module_with_0shape_data(test_case): testcase的自动测试条件设为 @autotest(auto_backward=False, check_graph=True)时报错，出错信息如下： [interface_op_util.cpp:28] Check failed: shape.At(i) > 0
![12 7-1](https://user-images.githubusercontent.com/52237830/145007992-cc7a2bd5-a915-4075-aa86-6a545cbe8c8a.png)
从打印出的日志确实出现了shape.At(i)=0的情况出现所以把interface_op_util.cpp中的CHECK_GT(shape.At(i), 0)改为CHECK_GE(shape.At(i), 0)后还会报Check failed: chunk->mem_size()>0错误，把 plan_util.cpp中的CHECK_GT(chunk->mem_size(), 0)注释掉后跑通了testcase
![12 7-2](https://user-images.githubusercontent.com/52237830/145009194-e12f367b-e10c-439b-8a44-d26b17e56516.jpg)



**计划
如果今天通过review后合并PR

grybd commented:
在跑test_activation.py测试脚本时把跑def test_relu_module_with_0shape_data(test_case): testcase的自动测试条件设为 @autotest(auto_backward=False, check_graph=True)时报错，出错信息如下： [interface_op_util.cpp:28] Check failed: shape.At(i) > 0
![12 7-1](https://user-images.githubusercontent.com/52237830/145007992-cc7a2bd5-a915-4075-aa86-6a545cbe8c8a.png)
从打印出的日志确实出现了shape.At(i)=0的情况出现所以把interface_op_util.cpp中的CHECK_GT(shape.At(i), 0)改为CHECK_GE(shape.At(i), 0)后还会报Check failed: chunk->mem_size()>0错误从打印的日志来看确实会出现有mem_size()=0的情况
![12 7-3](https://user-images.githubusercontent.com/52237830/145018075-b7ab80a8-c7f5-45b3-8733-e9b33908bbcd.png)
把 plan_util.cpp中的CHECK_GT(chunk->mem_size(), 0)注释掉后跑通了testcase
![12 7-2](https://user-images.githubusercontent.com/52237830/145009194-e12f367b-e10c-439b-8a44-d26b17e56516.jpg)

grybd commented:
通过了CI测试，现在走review流程

grybd commented:
#目前发现oneflow在运行过程中会出现0-D tensor的情况，因此做了些实验来验证什么情况下会出现0 shape tensor，以及哪些OP可以消费0 shape tensor
##第一种出现0 shape tensor 的情况是由于oneflow 对input tensor 执行 slice 和 sum()、ReduceSum、ShapeElemCnt 等OP而引入Scalar tensor ,此时shape=0,例子
```python
x = flow.randn(2)
print("x:",x)
print("x[0].shape:",x[0].shape)
print("x.sum():",x.sum(),"\n","x.sum() shape:",x.sum().shape,'\n')

output：
x: tensor([0.0034, 1.0565], dtype=oneflow.float32)
x[0].shape: oneflow.Size([])
x.sum(): tensor(1.0599, dtype=oneflow.float32) 
x.sum() shape: oneflow.Size([]) 
```
##当在eagr模式下由于对input tensor执行了slice 和 sum()等op使得执行forward()函数时就有可能把Scalar tensor往前传比如
```python
x = flow.tensor([1.0,2.0], device="cpu")
class SubModule(flow.nn.Module):
    def __init__(self):
        super().__init__()
        self.weight = flow.nn.Parameter(
            flow.tensor([1.0, 2.0, 3.0, 4.0], device="cpu")
        )
    def forward(self, x):
        print ((x * self.weight[3]).shape)

m = SubModule()
eager_out = SubModule(x[1]) #执行了slice,引入Scalar tensor

output:
loaded library: /lib64/libibverbs.so.1
oneflow.Size([])
```
##通过用户输入引入0-D tensor 比如下面这行代码 x = random_pytorch_tensor(4, 2, 3, 0, 3).to(device)会引入 0-D tensor ，与Scalar tensor不同它是一个为空的tensor，shape的也等于0，当引入这种tensor后跑TestReLUModule(flow.unittest.TestCase)就会报错，也就说ReLU OP 也会消费0-D tensor比如：
```python 
@flow.unittest.skip_unless_1n1d()
class TestReLUModule(flow.unittest.TestCase):
    @autotest(auto_backward=False, check_graph=True)
    def test_relu_module_with_0shape_data(test_case):
        m = torch.nn.ReLU()
        m.train(random())
        device = random_device()
        m.to(device)
        pdb.set_trace()
        x = random_pytorch_tensor(4, 2, 3, 0, 3).to(device)
        print(x)
        y = m(x)
        return y
if __name__ == "__main__":
    unittest.main()

output:
PyTorch object:
tensor([], device='cuda:0', size=(2, 3, 0, 3), grad_fn=<ToCopyBackward0>)
OneFlow object:
tensor([], device='cuda:0', size=(2, 3, 0, 3), dtype=oneflow.float32, grad_fn=<copy_backward>)
Check failed: shape.At(i) > 0 (0 vs. 0)
```

strint commented on python/oneflow/test/modules/test_activation.py:
```diff
@@ -39,7 +38,7 @@ def test_relu_module_with_random_data(test_case):
         return y
 
-    @autotest(auto_backward=False, check_graph=False)
+    @autotest(auto_backward=False, check_graph=True)
```
这里可以直接去掉check_graph=True，就能触发graph的执行0shape， 因为默认值为true。

grybd replied:
好的

grybd commented:
直接去掉check_graph=True，就能触发graph的执行0shape

grybd commented:
CI失败，现在重跑

grybd commented:
cuda 相关test fail 重跑CI

grybd commented:
## oneflow在运行过程中会出现0 shape tensor从而导致程序运行出错
为了解0 shape tensor 在系统中是如何流转的，便于后续开发支持0 shape tensor的情况，所以做实验验证什么情况下会出现0 shape tensor，以及哪些OP可以消费0 shape tensor，

### 第一种引入0 shape 的方式
通过index的方式引入 0 shape tensor：
```python
def test_relu_module_with_0shape_data(test_case):
        m = torch.nn.ReLU()
        m.train(random())
        device = random_device()
        m.to(device)
        x = random_pytorch_tensor(4, 2, 3, 0, 3).to(device)
        print("x:",x)
        print("x shape:",x.shape)
        print("x[1][2].shape",x[1][2].shape)
        print("input:",x[1][2])
        y = m(x)
        print("after ReLU:",y)
        return y
```


输出结果：
```
x: 
PyTorch object:  tensor([], size=(2, 3, 0, 3), requires_grad=True)
OneFlow object:  tensor([], size=(2, 3, 0, 3), dtype=oneflow.float32, requires_grad=True)

x shape: 
PyTorch object: torch.Size([2, 3, 0, 3])
OneFlow object:  oneflow.Size([2, 3, 0, 3])

x[1][2].shape
PyTorch object:   torch.Size([0, 3])
OneFlow object:  oneflow.Size([0, 3])

input:
PyTorch object:    tensor([], size=(0, 3), grad_fn=<SelectBackward0>)
OneFlow object:   tensor([], size=(0, 3), dtype=oneflow.float32, grad_fn=<view::reshape_backward>)

after ReLU: 
PyTorch object:
tensor([], size=(2, 3, 0, 3), grad_fn=<ReluBackward0>)
OneFlow object:
tensor([], size=(2, 3, 0, 3), dtype=oneflow.float32, grad_fn=<relu_backward>)
```
### 第二种引入0 shape 的方式
通过slice方式引入0 shape tensor
```python
def test_relu_module_with_0shape_data(test_case):
        m = torch.nn.ReLU()
        m.train(random())
        device = random_device()
        m.to(device)
        x = random_pytorch_tensor(4, 2, 3, 0, 3).to(device)
        print("x:",x)
        print("x.shape:",x.shape)
        print("x[0:0,:].shape",x[0:0,:].shape)
        print("input:",x[0:0,:])
        y = m(x[0:0,:])
        print("after ReLU:",y)
        return y
```


输出结果:
```
x: 
 PyTorch object:  tensor([], device='cuda:0', size=(2, 3, 0, 3), grad_fn=<ToCopyBackward0>)
 OneFlow object:  tensor([], device='cuda:0', size=(2, 3, 0, 3), dtype=oneflow.float32, grad_fn=<copy_backward>)

x.shape:
 PyTorch object:  torch.Size([2, 3, 0, 3])
 OneFlow object:  oneflow.Size([2, 3, 0, 3])

x[0:0,:].shape
 PyTorch object:  torch.Size([0, 3, 0, 3])
 OneFlow object:  oneflow.Size([0, 3, 0, 3])

input: 
 PyTorch object:  tensor([], device='cuda:0', size=(0, 3, 0, 3), grad_fn=<SliceBackward0>)
 OneFlow object:  tensor([], device='cuda:0', size=(0, 3, 0, 3), dtype=oneflow.float32, grad_fn=<slice_backward>)

after ReLU: 
PyTorch object:  tensor([], device='cuda:0', size=(0, 3, 0, 3), grad_fn=<ReluBackward0>)
OneFlow object:   tensor([], device='cuda:0', size=(0, 3, 0, 3), dtype=oneflow.float32, grad_fn=<relu_backward>)
```
### 第三种用户引入 0 shape tensor
通过用户引入0 shape tensor 比如下面这行代码 x = random_pytorch_tensor(4, 2, 3, 0, 3).to(device)会引入 0 shape tensor ，与Scalar tensor不同它是一个为空的tensor，shape等于0，由于graph还未支持0 shape tensor 所以当引入这种tensor后跑TestReLUModule(flow.unittest.TestCase)，会报错，见代码：
```python 
@flow.unittest.skip_unless_1n1d()
class TestReLUModule(flow.unittest.TestCase):
    @autotest(auto_backward=False, check_graph=True)
    def test_relu_module_with_0shape_data(test_case):
        m = torch.nn.ReLU()
        m.train(random())
        device = random_device()
        m.to(device)
        pdb.set_trace()
        x = random_pytorch_tensor(4, 2, 3, 0, 3).to(device)
        print(x)
        y = m(x)
        return y
if __name__ == "__main__":
    unittest.main()
```


输出结果:
```
PyTorch object:
tensor([], device='cuda:0', size=(2, 3, 0, 3), grad_fn=<ToCopyBackward0>)
OneFlow object:
tensor([], device='cuda:0', size=(2, 3, 0, 3), dtype=oneflow.float32, grad_fn=<copy_backward>)
Check failed: shape.At(i) > 0 (0 vs. 0)
```

doombeaker commented:
我发现一个小问题

![image](https://user-images.githubusercontent.com/3351623/146875335-2d4d246a-a7d6-4a1b-bdc0-16d265b73db2.png)

为什么这里的提交者，全部都是 one-bot。不知道是不是提交时 email 没有设置对

grybd commented:
> 我发现一个小问题，这个PR里的commmit
> 
> ![image](https://user-images.githubusercontent.com/3351623/146875335-2d4d246a-a7d6-4a1b-bdc0-16d265b73db2.png)
> 
> 全部都是 one-bot。不知道是不是提交时 email 没有设置对

我检查一下

grybd commented:
> 我发现一个小问题，这个PR里的commmit
> 
> ![image](https://user-images.githubusercontent.com/3351623/146875335-2d4d246a-a7d6-4a1b-bdc0-16d265b73db2.png)
> 
> 全部都是 one-bot。不知道是不是提交时 email 没有设置对

确实是，切换了目录了，忘记设置邮箱和用户名了

grybd commented:
设置了邮箱和用户名，merge master into this branch 现在进行CI

grybd commented:
## 0 shape tensor 在relu_op 和 relu_kernel中的处理流程
以relu op处理0 shape tensor的情况为例，在relu_op 和 relu_kernel中打印日志，了解0 shape tensor 在系统中是如何流转的
### 第一步跑一个与0 shape tensor 相关的testcase
通过追踪这个testcase 了解0 shape tensor 在relu_op 和 relu_kernel中的处理流程
@flow.unittest.skip_unless_1n1d()
class TestReLUModule(flow.unittest.TestCase):
    '''
    @autotest()
    def test_relu_module_with_random_data(test_case):
        m = torch.nn.ReLU()
        m.train(random())
        device = random_device()
        m.to(device)
        x = random_pytorch_tensor().to(device)
        y = m(x)
        return y
'''
    @autotest(auto_backward=False)
    def test_relu_module_with_0shape_data(test_case):
        print("test_relu_module_with_0shape_data(test_case):")
        m = torch.nn.ReLU()
        m.train(random())
        device = random_device()
        m.to(device)
        x = random_pytorch_tensor(4, 2, 3, 0, 3).to(device)
        print("x:",x)
        y = m(x)
        print("y:",y)
        return y

跑上述testcase后的动作是第一步oneflow会通过宏REGISTER_USER_OP 注册relu op 返回一个OpRegistry& 对象 再通过链式初始化，通过SetTensorDescInferFn([](user_op::InferContext* ctx)函数推到出 input shape 和 output shape等逻辑信息
REGISTER_USER_OP("relu")
    .Input("in")
    .Output("out")
    .SetTensorDescInferFn([](user_op::InferContext* ctx) -> Maybe<void> {
      const Shape& in_shape = ctx->InputShape("in", 0);
      //std::cout<<"in_shape:"<< in_shape.elem_cnt()<<std::endl;
      Shape* out_shape = ctx->OutputShape("out", 0);
      //std::cout<<"out_shape:"<< out_shape->elem_cnt()<<std::endl;
      *out_shape = in_shape;
      return Maybe<void>::Ok();
    })
    .SetGetSbpFn([](user_op::SbpContext* ctx) -> Maybe<void> {
      const user_op::TensorDesc& in_tensor = ctx->LogicalTensorDesc4InputArgNameAndIndex("in", 0);
      FOR_RANGE(int64_t, i, 0, in_tensor.shape().NumAxes()) {
        ctx->NewBuilder()
            .Split(user_op::OpArg("in", 0), i)
            .Split(user_op::OpArg("out", 0), i)
            .Build();
      }
      return Maybe<void>::Ok();
    })
    .SetDataTypeInferFn([](user_op::InferContext* ctx) -> Maybe<void> {
      *ctx->OutputDType("out", 0) = ctx->InputDType("in", 0);
      return Maybe<void>::Ok();
    });

然后通过relu_kernel 执行具体的计算逻辑返回计算结果，在relu_kernel中通过注册的保存在KernelComputeContext逻辑信息，调用 Compute方法计算得出结果，Compute方法里会判断 input tensor的elem_cnt，如果不等于0会调用primitive->Launch(ctx->stream(), x->dptr(), y->mut_dptr(), elem_cnt);将结果返回，如果等于0 即是0 shape tensor即什么都不干直接返回[]，具体的处理逻辑代码如下：
 void Compute(user_op::KernelComputeContext* ctx) const override {
    auto primitive = NewReluPrimitive(ctx);
    CHECK(primitive);

    const user_op::Tensor* x = ctx->Tensor4ArgNameAndIndex("in", 0);
    //std::cout<<"y elem_cnt:"<<x->shape().elem_cnt()<<"x NumAxes:" << x->shape().NumAxes()<<std::endl;
    user_op::Tensor* y = ctx->Tensor4ArgNameAndIndex("out", 0);
    //std::cout<<"y elem_cnt:"<<y->shape().elem_cnt()<<"y NumAxes:" << y->shape().NumAxes()<<std::endl;
    const int64_t elem_cnt = x->shape().elem_cnt();

    if (elem_cnt != 0) {
      primitive->Launch(ctx->stream(), x->dptr(), y->mut_dptr(), elem_cnt);
    } else {
      // For 0-shape Tensor
      std::cout<< "0-shape Tensor" << std::endl;
      return;
    }

doombeaker commented:
这个PR解决什么问题，能否在 PR 描述了提供下复现问题的最小代码。

grybd commented:
> 这个PR解决什么问题，能否在 PR 描述了提供下复现问题的最小代码。

已经添加PR描述

strint commented:
> ## 0 shape tensor 在relu_op 和 relu_kernel中的处理流程
> 然后通过relu_kernel 执行具体的计算逻辑返回计算结果，在relu_kernel中通过前面在relu op中注册时保存在KernelComputeContext里面的逻辑信息，调用 Compute方法计算得出结果，Compute方法里会判断 input tensor的elem_cnt，如果不等于0会调用primitive->Launch(ctx->stream(), x->dptr(), y->mut_dptr(), elem_cnt);将结果返回，如果等于0 即是0 shape tensor即什么都不做直接返回[]，具体的处理逻辑代码如下：
> 
> ```c++
>  void Compute(user_op::KernelComputeContext* ctx) const override {
>     auto primitive = NewReluPrimitive(ctx);
>     CHECK(primitive);
> 
>     const user_op::Tensor* x = ctx->Tensor4ArgNameAndIndex("in", 0);
>     //std::cout<<"x elem_cnt:"<<x->shape().elem_cnt()<<"x NumAxes:" << x->shape().NumAxes()<<std::endl;
>     user_op::Tensor* y = ctx->Tensor4ArgNameAndIndex("out", 0);
>     //std::cout<<"y elem_cnt:"<<y->shape().elem_cnt()<<"y NumAxes:" << y->shape().NumAxes()<<std::endl;
>     const int64_t elem_cnt = x->shape().elem_cnt();
> 
>     if (elem_cnt != 0) {
>       primitive->Launch(ctx->stream(), x->dptr(), y->mut_dptr(), elem_cnt);
>     } else {
>       // For 0-shape Tensor
>       std::cout<< "0-shape Tensor" << std::endl;
>       return;
>     }
> ```

没问题，可以处理剩余的0 shape case了。

grybd commented:
> > ## 0 shape tensor 在relu_op 和 relu_kernel中的处理流程
> > 然后通过relu_kernel 执行具体的计算逻辑返回计算结果，在relu_kernel中通过前面在relu op中注册时保存在KernelComputeContext里面的逻辑信息，调用 Compute方法计算得出结果，Compute方法里会判断 input tensor的elem_cnt，如果不等于0会调用primitive->Launch(ctx->stream(), x->dptr(), y->mut_dptr(), elem_cnt);将结果返回，如果等于0 即是0 shape tensor即什么都不做直接返回[]，具体的处理逻辑代码如下：
> > ```c++
> >  void Compute(user_op::KernelComputeContext* ctx) const override {
> >     auto primitive = NewReluPrimitive(ctx);
> >     CHECK(primitive);
> > 
> >     const user_op::Tensor* x = ctx->Tensor4ArgNameAndIndex("in", 0);
> >     //std::cout<<"x elem_cnt:"<<x->shape().elem_cnt()<<"x NumAxes:" << x->shape().NumAxes()<<std::endl;
> >     user_op::Tensor* y = ctx->Tensor4ArgNameAndIndex("out", 0);
> >     //std::cout<<"y elem_cnt:"<<y->shape().elem_cnt()<<"y NumAxes:" << y->shape().NumAxes()<<std::endl;
> >     const int64_t elem_cnt = x->shape().elem_cnt();
> > 
> >     if (elem_cnt != 0) {
> >       primitive->Launch(ctx->stream(), x->dptr(), y->mut_dptr(), elem_cnt);
> >     } else {
> >       // For 0-shape Tensor
> >       std::cout<< "0-shape Tensor" << std::endl;
> >       return;
> >     }
> > ```
> 
> 没问题，可以处理剩余的0 shape case了。

好的，CI通过了，你们有空review一下

strint commented:
### TODO
1、改名 0shape -> 0_size

2、https://github.com/Oneflow-Inc/OneTeam/issues/899 中通过不了的case的说明，具体到文件 + case名

之后就合并该pr

grybd commented:
> ### TODO
> 1、改名 0shape -> 0_size
> 
> 2、[Oneflow-Inc/OneTeam#899](https://github.com/Oneflow-Inc/OneTeam/issues/899) 中通过不了的case的说明，具体到文件 + case名
> 
> 之后就合并该pr
> 
> 今天需要完成

好的

strint commented on oneflow/core/job/plan_util.cpp:
```diff
@@ -163,7 +163,7 @@ void GenChunkForMultiNNGraphMemoryReuseInMultiClient(
           CHECK_GE(current_chunk_offset, 0);
           // CHECK_GT(mem_block->mem_size(), 0); NOTE(chengcheng): has mem block mem size = 0
-          CHECK_GT(chunk->mem_size(), 0);
+          CHECK_GE(chunk->mem_size(), 0);
```
这里改动少，但是代码层次很深。也请 @chengtbf review下

chengtbf replied:
这里可以这样改。之前觉得不可能一个 device 上不申请任何显存。但这个 case 里就遇到了。整个 chunk 都是 null，那么实际上也不会给这个 chunk 分配显存。  runtime 不需要做改动吗？

strint replied:
是不需要改动，就可以跑通。

已经通过过CI。

strint commented on python/oneflow/test/modules/test_sub.py:
```diff
@@ -127,8 +127,8 @@ def test_sub_against_pytorch(test_case):
-    @autotest(auto_backward=False, check_graph=False)
-    def test_sub_with_0shape_data(test_case):
+    @autotest(auto_backward=False, check_graph=True)
+    def test_sub_with_0_size_data(test_case):
```
现在ci这个case报错了
```
[/opt/conda/bin/python3 /home/ci-user/runners/gpu/_work/oneflow/oneflow/test_tmp_dir/modules/test_sub.py --failfast --verbose] F1229 16:28:38.494190 25141 intra_job_mem_sharing_util.cpp:447] Check failed: begin < end && end <= buffer_size_ 
[/opt/conda/bin/python3 /home/ci-user/runners/gpu/_work/oneflow/oneflow/test_tmp_dir/modules/test_sub.py --failfast --verbose] *** Check failure stack trace: ***
```

这个case不是 https://github.com/Oneflow-Inc/oneflow/pull/7142 后来解决的么

这里不能该的

grybd replied:
> 现在ci这个case报错了
> 
> ```
> [/opt/conda/bin/python3 /home/ci-user/runners/gpu/_work/oneflow/oneflow/test_tmp_dir/modules/test_sub.py --failfast --verbose] F1229 16:28:38.494190 25141 intra_job_mem_sharing_util.cpp:447] Check failed: begin < end && end <= buffer_size_ 
> [/opt/conda/bin/python3 /home/ci-user/runners/gpu/_work/oneflow/oneflow/test_tmp_dir/modules/test_sub.py --failfast --verbose] *** Check failure stack trace: ***
> ```
> 
> 这个case不是 #7142 后来解决的么
> 
> 这里不能改的

已处理

strint commented on python/oneflow/test/modules/test_add.py:
```diff
@@ -183,15 +183,15 @@ def test_0dim_inplace_add(test_case):
         x = random_pytorch_tensor(2, 2, 3).to(device).mean()
         y = random_pytorch_tensor(2, 2, 3).to(device)
-        x += y.mean()
+        x = x + y.mean()
```
这个是不能改的，这里本来就是测试inplace的

grybd replied:
这个好像是format改过来的，我改回去

strint commented on python/oneflow/test/modules/test_add.py:
```diff
@@ -183,15 +183,15 @@ def test_0dim_inplace_add(test_case):
         x = random_pytorch_tensor(2, 2, 3, requires_grad=False).to(device)
         y = random_pytorch_tensor(1, 10).to(device)
-        x += y.mean()
+        x = x + y.mean()
```
这里也不能改，也是测试inplace的。并没有check_graph，按说也不用改。

strint commented on python/oneflow/test/modules/test_greater.py:
```diff
@@ -119,7 +119,7 @@ def test_tensor_greater_with_random_data(test_case):
 
-    @autotest(auto_backward=False, check_graph=False)
+    @autotest(auto_backward=False, check_graph=True)
     def test_greater_with_0shape_data(test_case):
```
这些本来改成0 size，怎么又改回0 shape了？

grybd replied:
是0 size 的，应该是合并了master分支忘记改过来了，这个我没注意到，我现在给改正

strint commented:
CI报了这个case的错误
```
@autotest(check_graph=True)
    def test_flow_var_one_dim_with_random_data(test_case):
        device = random_device()
        x = random_pytorch_tensor(ndim=4).to(device)
        y = torch.var(
            x,
            dim=random(low=0, high=4).to(int),
            unbiased=random().to(bool),
            keepdim=random().to(bool),
        )
        return y
```
错误如下
```
test_flow_var_one_dim_with_random_data (test_var.TestVar) ... FAIL

======================================================================
FAIL: test_flow_var_one_dim_with_random_data (test_var.TestVar)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/site-packages/oneflow/test_utils/automated_test_util/torch_flow_dual_object.py", line 709, in new_f
    equal_nan=True,
AssertionError: False is not true

```

看这个case的确是新打开的 chech_graph = True。
@grybd 本地试验下看能否复现

grybd commented:
> CI报了这个case的错误
> 
> ```
> @autotest(check_graph=True)
>     def test_flow_var_one_dim_with_random_data(test_case):
>         device = random_device()
>         x = random_pytorch_tensor(ndim=4).to(device)
>         y = torch.var(
>             x,
>             dim=random(low=0, high=4).to(int),
>             unbiased=random().to(bool),
>             keepdim=random().to(bool),
>         )
>         return y
> ```
> 
> 错误如下
> 
> ```
> test_flow_var_one_dim_with_random_data (test_var.TestVar) ... FAIL
> 
> ======================================================================
> FAIL: test_flow_var_one_dim_with_random_data (test_var.TestVar)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/opt/conda/lib/python3.7/site-packages/oneflow/test_utils/automated_test_util/torch_flow_dual_object.py", line 709, in new_f
>     equal_nan=True,
> AssertionError: False is not true
> ```
> 
> 看这个case的确是新打开的 chech_graph = True。 @grybd 本地试验下看能否复现

这个试过了，单测是可以通过的

strint commented:
> > CI报了这个case的错误
> > ```
> > @autotest(check_graph=True)
> >     def test_flow_var_one_dim_with_random_data(test_case):
> >         device = random_device()
> >         x = random_pytorch_tensor(ndim=4).to(device)
> >         y = torch.var(
> >             x,
> >             dim=random(low=0, high=4).to(int),
> >             unbiased=random().to(bool),
> >             keepdim=random().to(bool),
> >         )
> >         return y
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 错误如下
> > ```
> > test_flow_var_one_dim_with_random_data (test_var.TestVar) ... FAIL
> > 
> > ======================================================================
> > FAIL: test_flow_var_one_dim_with_random_data (test_var.TestVar)
> > ----------------------------------------------------------------------
> > Traceback (most recent call last):
> >   File "/opt/conda/lib/python3.7/site-packages/oneflow/test_utils/automated_test_util/torch_flow_dual_object.py", line 709, in new_f
> >     equal_nan=True,
> > AssertionError: False is not true
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 看这个case的确是新打开的 chech_graph = True。 @grybd 本地试验下看能否复现
> 
> 这个试过了，单测是可以通过的

我还是先去掉了。你本地把autotest(n）里面的n调大，看看能不能复现。

grybd commented:
> > > CI报了这个case的错误
> > > ```
> > > @autotest(check_graph=True)
> > >     def test_flow_var_one_dim_with_random_data(test_case):
> > >         device = random_device()
> > >         x = random_pytorch_tensor(ndim=4).to(device)
> > >         y = torch.var(
> > >             x,
> > >             dim=random(low=0, high=4).to(int),
> > >             unbiased=random().to(bool),
> > >             keepdim=random().to(bool),
> > >         )
> > >         return y
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > 错误如下
> > > ```
> > > test_flow_var_one_dim_with_random_data (test_var.TestVar) ... FAIL
> > > 
> > > ======================================================================
> > > FAIL: test_flow_var_one_dim_with_random_data (test_var.TestVar)
> > > ----------------------------------------------------------------------
> > > Traceback (most recent call last):
> > >   File "/opt/conda/lib/python3.7/site-packages/oneflow/test_utils/automated_test_util/torch_flow_dual_object.py", line 709, in new_f
> > >     equal_nan=True,
> > > AssertionError: False is not true
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > 看这个case的确是新打开的 chech_graph = True。 @grybd 本地试验下看能否复现
> > 
> > 
> > 这个试过了，单测是可以通过的
> 
> 我还是先去掉了。你本地把autotest(n）里面的n调大，看看能不能复现。

好的
